src/hooks/useGasLimit.ts

```diff
@@ -68,6 +148,11 @@ const useGasLimit = (
       safeTx.signatures.size < threshold,
     )
 
+    // if we are dealing with zksync and the walletAddress is a Safe, we have to do some magic
+    if (safe.chainId === chains.zksync && (await web3ReadOnly.getCode(walletAddress)) !== '0x') {
+      return getGasLimitForZkSync(safe, web3ReadOnly, safeSDK, safeTx)
+    }
```

Comment:
Why do we have to reimplement the entire gas estimation and not just substitute the [from address](https://github.com/safe-global/safe-wallet-web/blob/32f9038386419ec31d65dd06eee64d4340ba53b5/src/hooks/useGasLimit.ts#L159) in the existing estimation call?

Also, there's already [a multiplier for Gnosis Chain](https://github.com/safe-global/safe-wallet-web/blob/32f9038386419ec31d65dd06eee64d4340ba53b5/src/hooks/useGasLimit.ts#L165-L168), it would be good to do it for zkSync in the same place, using some config.

Reply:
The problem here is:
On Zksync every EOA is actually a Smart account deployed through the same proxy factory.
You can only use `eth_estimateGas` if the `from` address was created through that proxy factory. Safes do not use that factory and thus can not be used.

If we now put a random EOA as `from` address and call `eth_estimateGas` the transaction will fail as the signature is missing because that random EOA is not an owner and therefore we can not use a prevalidated signature.
That's why we have to use this alternative estimation function. 

@compojoom requested to add all this logic to the core-sdk. That would be the ideal solution as this affects all dapps built on top of Safe using the protocol-kit.

Reply:
Because it's not that easy :) If we just change the `from` address we run into'GS025: Hash has not been approved' (https://github.com/safe-global/safe-smart-account/blob/12136fd54940892bf43258f86b6e8f9d0eff69ea/contracts/Safe.sol#L302). If we also change the `from` also for the enoded tx we get`GS026: Invalid owner provided`. 

So, instead what we do for zksync is we use the simulate function on the callbackHandler: https://github.com/safe-global/safe-smart-account/blob/12136fd54940892bf43258f86b6e8f9d0eff69ea/contracts/handler/CompatibilityFallbackHandler.sol#L88
So we are not passing the encodedSafeTx for the estimateGas function, but rather "wrapping" it in a "simulate" call and then calling the estimate gas for "simulate" call. 

That's my current understanding of the issue. @schmanu did the same thing for counterfactual deployments, but there he had a multisend. 

About the multiplier - the multiplier for GNO is not a config. Also the incrementByPercentage function just increases the gasLimit by that percentage. Note that I'm multiplying the baseFee by 20 to get it to a value that passes.  So, should I use incrementByPercentage and pass 2000%?

Reply:
@compojoom I would stick to floats, i.e.

```js
const GasMultipliers = {
  [chains.gno]: 1.2,
  [chains.zksync]: 20,
}
```

Reply:
why float and not Bigint? The returned values from the estimate functions are bigints